### PR TITLE
AMDGPU/NewPM Port GCNDPPCombine to NPM

### PR DESCRIPTION
**Pull request by optimisan**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp [hunk 1]
```diff
@@ -76,12 +77,17 @@ class GCNDPPCombine : public MachineFunctionPass {
 
   bool combineDPPMov(MachineInstr &MI) const;
 
+  int getDPPOp(unsigned Op, bool IsShrinkable) const;
+  bool isShrinkable(MachineInstr &MI) const;
+
+public:
+  bool run(MachineFunction &MF);
+};
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:85]: Missing getRequiredProperties() in the NPM path.

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp [hunk 2]
```diff
@@ -749,9 +752,17 @@ bool GCNDPPCombine::combineDPPMov(MachineInstr &MovMI) const {
   return !Rollback;
 }
 
-bool GCNDPPCombine::runOnMachineFunction(MachineFunction &MF) {
+bool GCNDPPCombineLegacy::runOnMachineFunction(MachineFunction &MF) {
+  if (skipFunction(MF.getFunction())) {
+    return false;
+  }
+  GCNDPPCombine Impl;
+  return Impl.run(MF);
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:760]:
```suggestion
  return  GCNDPPCombine().run(MF);
```

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp [hunk 3]
```diff
@@ -781,3 +792,15 @@ bool GCNDPPCombine::runOnMachineFunction(MachineFunction &MF) {
   }
   return Changed;
 }
+
+PreservedAnalyses GCNDPPCombinePass::run(MachineFunction &MF,
+                                         MachineFunctionAnalysisManager &) {
+  GCNDPPCombine Impl;
+  bool Changed = Impl.run(MF);
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:799]:
```suggestion
  bool Changed = GCNDPPCombine().run(MF);
```

File: llvm/test/CodeGen/AMDGPU/dpp64_combine.mir [hunk 4]
```diff
@@ -1,5 +1,5 @@
-# RUN: llc -mtriple=amdgcn -mcpu=gfx90a -run-pass=gcn-dpp-combine -verify-machineinstrs -o - %s | FileCheck %s --check-prefix=GCN
-# RUN: llc -mtriple=amdgcn -mcpu=gfx940 -run-pass=gcn-dpp-combine -verify-machineinstrs -o - %s | FileCheck %s --check-prefix=GCN
+# RUN: llc -mtriple=amdgcn -mcpu=gfx90a -passes=gcn-dpp-combine -verify-machineinstrs -o - %s | FileCheck %s --check-prefix=GCN
```
Comment (cdevadas) [inline llvm/test/CodeGen/AMDGPU/dpp64_combine.mir:1]: Retain the test behavior for legacy pm. Either duplicate an additional line for NPM or add no change. Rectify the same for other tests below.

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+//    $bound_ctrl==DPP_BOUND_OFF and
+//    $old==identity value (immediate) for the VALU op
+// -> $combined_old = src1,
+//    $combined_bound_ctrl = DPP_BOUND_OFF
+//
+// Otherwise cancel.
+//
+// The mov_dpp instruction should reside in the same BB as all its uses
+//===----------------------------------------------------------------------===//
+
```
Comment (pravinjagtap) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.h:40]: This is same as llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp description. Avoid the duplication.

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp [hunk 6]
```diff
@@ -76,12 +77,17 @@ class GCNDPPCombine : public MachineFunctionPass {
 
   bool combineDPPMov(MachineInstr &MI) const;
 
+  int getDPPOp(unsigned Op, bool IsShrinkable) const;
+  bool isShrinkable(MachineInstr &MI) const;
+
+public:
+  bool run(MachineFunction &MF);
+};
+class GCNDPPCombineLegacy : public MachineFunctionPass {
```
Comment (pravinjagtap) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:86]: Move to a next line

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp [hunk 7]
```diff
@@ -749,9 +753,16 @@ bool GCNDPPCombine::combineDPPMov(MachineInstr &MovMI) const {
   return !Rollback;
 }
 
-bool GCNDPPCombine::runOnMachineFunction(MachineFunction &MF) {
+bool GCNDPPCombineLegacy::runOnMachineFunction(MachineFunction &MF) {
+  if (skipFunction(MF.getFunction())) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:757]: Don't need braces

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp [hunk 8]
```diff
@@ -781,3 +792,14 @@ bool GCNDPPCombine::runOnMachineFunction(MachineFunction &MF) {
   }
   return Changed;
 }
+
+PreservedAnalyses GCNDPPCombinePass::run(MachineFunction &MF,
+                                         MachineFunctionAnalysisManager &) {
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:797]: Add the following to skip this pass for OptNone cases.
if (MF.getFunction().hasOptNone())
   return false;

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp [hunk 9]
```diff
@@ -781,3 +792,14 @@ bool GCNDPPCombine::runOnMachineFunction(MachineFunction &MF) {
...
+  bool Changed = GCNDPPCombine().run(MF);
+  if (!Changed) {
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:799]: No braces needed.

File: llvm/lib/Target/AMDGPU/GCNDPPCombine.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  MachineFunctionProperties getRequiredProperties() {
+    return MachineFunctionProperties().set(
+        MachineFunctionProperties::Property::IsSSA);
+  }
+};
+
+} // end namespace llvm
+
+#endif
```
[the remark this answers is not in the document]
Comment (Akshat-Oke) [reply llvm/lib/Target/AMDGPU/GCNDPPCombine.h:28]: `FixNamespaceComments` adds it for namespaces but there's probably nothing for `#endif`s.

**Review by cdevadas — APPROVED** [review]

**Pull request merged**